src/circuit/gadget/ecc.rs

```diff
@@ -1,140 +1,336 @@
-    type Point: Clone + fmt::Debug;
+    type Point: Clone + Debug;
+    /// Variable representing the x-coordinate of an elliptic curve point.
+    type X: Clone + Debug;
     /// Variable representing the set of fixed bases in the circuit.
-    type FixedPoints: FixedPoints<C>;
+    type FixedPoints: Clone + Debug;
     /// Variable representing a fixed elliptic curve point (constant in the circuit).
-    type FixedPoint: Clone + fmt::Debug;
+    type FixedPoint: Clone + Debug;
+    /// Variable representing a fixed elliptic curve point (constant in the circuit)
+    /// to be used in scalar multiplication with a short signed exponent.
+    type FixedPointShort: Clone + Debug;
+
+    /// Witnesses the given base field element as a private input to the circuit for variable-base scalar mul.
+    fn witness_scalar_var(
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
+        value: Option<C::Base>,
+    ) -> Result<Self::ScalarVar, Error>;
 
-    /// Witnesses the given scalar as a private input to the circuit.
-    fn witness_scalar(
-        layouter: &mut impl Layouter<Self>,
+    /// Witnesses the given full-width scalar as a private input to the circuit for fixed-base scalar mul.
+    fn witness_scalar_fixed(
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
         value: Option<C::Scalar>,
-    ) -> Result<Self::Scalar, Error>;
+    ) -> Result<Self::ScalarFixed, Error>;
+
+    /// Witnesses the given signed short scalar as a private input to the circuit for fixed-base scalar mul.
+    fn witness_scalar_fixed_short(
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
+        value: Option<C::Scalar>,
+    ) -> Result<Self::ScalarFixedShort, Error>;
 
     /// Witnesses the given point as a private input to the circuit.
     fn witness_point(
-        layouter: &mut impl Layouter<Self>,
+        &self,
+        layouter: &mut impl Layouter<C::Base>,
         value: Option<C>,
     ) -> Result<Self::Point, Error>;
 
-    /// Gets a fixed point into the circuit.
-    fn get_fixed(
-        layouter: &mut impl Layouter<Self>,
+    /// Extracts the x-coordinate of a point.
+    fn extract_p(point: &Self::Point) -> &Self::X;
+
+    /// Returns a fixed point that had been previously loaded into the circuit.
+    /// The pre-loaded cells are used to set up equality constraints in other
+    /// parts of the circuit where the fixed base is used.
+    fn get_fixed(&self, fixed_points: Self::FixedPoints) -> Result<Self::FixedPoint, Error>;
+
+    /// Returns a fixed point to be used in scalar multiplication with a signed
+    /// short exponent.
+    fn get_fixed_short(
+        &self,
         fixed_points: Self::FixedPoints,
-    ) -> Result<Self::FixedPoint, Error>;
+    ) -> Result<Self::FixedPointShort, Error>;
 
-    /// Performs point addition, returning `a + b`.
-    fn add(
-        layouter: &mut impl Layouter<Self>,
+    /// Performs incomplete point addition, returning `a + b`.
```

Comment:
Document the expected behaviour on exceptional cases (either "always returns an error", or "deterministic but undefined behaviour"). We could propagate the bottom case out of this API, but that would require tracking it in the circuit which we don't want to do for performance.

Reply:
My preferred behaviour would be to return Error when incomplete addition is defined to return ⊥. (The prover can always check that even if the circuit doesn't.)